activemodel/lib/active_model/attribute_methods.rb

```diff
@@ -336,7 +358,37 @@ def undefine_attribute_methods
       end
 
       private
-        class CodeGenerator
+        class CodeGenerator # :nodoc:
+          class MethodSet
+            METHOD_CACHES = Hash.new { |h, k| h[k] = Module.new }
+
+            def initialize(namespace)
+              @cache = METHOD_CACHES[namespace]
```

Comment:
It seems like a `MethodSet` instance should only be concerned with its own "cache" module, i.e.:

```suggestion
            def initialize
              @cache = Module.new
```

And then a `CodeGenerator::METHOD_SETS` Hash would replace the `CodeGenerator` `@namespaces` instance variable.  But I suppose you chose the current way to reduce retained memory (i.e. only hold on to the `Module` instances instead of the `MethodSet` instances)?

Reply:
It sounds this way but no. It's important that the cache modules are global. Because the idea is that both `Model1` and `Model2` have methods in common (typically `updated_at`, `updated_at_was`, etc).

So `Model1` and `Model2` must both first define the methods in a shared cache module, and then copy it from there.